builtin/game/hud.lua

```diff
@@ -0,0 +1,225 @@
+--[[
+Register function to easily register new builtin hud elements
+`def` is a table and contains the following fields:
+  elem_def  the HUD element definition which can be changed with hud_replace_builtin
+  event     (optional) for an additional eventname on which the element will be updated
+            ("hud_changed" and "properties_changed" will always be used.)
+  hud_change(id, player)
+            (optional) a function to change the element after it has been updated
+  show_elem(player, flags, id)
+            (optional) a function to decide if the element should be shown to a player
+            It is called before the element gets updated.
+]]--
+local registered_elements = {}
+local update_events = {}
+local function register_builtin_hud_element(name, def)
+	registered_elements[name] = def
+	if def.event then
+		update_events[def.event] = update_events[def.event] or {}
+		table.insert(update_events[def.event], name)
+	end
+end
+
+-- Stores HUD ids for all players
+local hud_ids = {}
+
+-- Updates one element
+-- In case the element is already added, it does only call the hud_change function from
+-- registered_elements. (To completely update the element remove it first.)
+local function update_element(player, player_hud_ids, elem_name, flags)
+	local def = registered_elements[elem_name]
+	local id = player_hud_ids[elem_name]
+
+	if def.show_elem and not def.show_elem(player, flags, id) then
+		if id then
+			player:hud_remove(id)
+			player_hud_ids[elem_name] = nil
+		end
+		return
+	end
+
+	if not id then
+		id = player:hud_add(def.elem_def)
+		player_hud_ids[elem_name] = id
+	end
+
+	if def.hud_change then
+		def.hud_change(player, id)
+	end
```

Comment:
> it should be easy enough to fix by providing elem_def as a function returning a table.

This is not possible, since `elem_def` is set by `hud_replace_builtin`.

Instead, I added a `init_change` function. Now it looks like code duplication, but at least it is faster.